### ISSUE TITLE
Start detached container

### ISSUE THREAD

**Issue by mastertinner** [bfirsh/whalebrew#53]
I would like to create a whalebrew package for [`mongod`](https://docs.mongodb.com/manual/reference/program/mongod/) which starts MongoDB as a daemon. I think it would be awesome if I could just run `mongodb` and whalebrew would start my MongoDB in a detached Docker container. As far as I can tell, all we need for that is a Docker image from the following `Dockerfile`:

```Dockerfile
FROM mongo

EXPOSE 27017

CMD mongod
```

And a new config label for whalebrew called `io.whalebrew.config.detach` which would set the `-d` flag when running the container. What do you guys think?

**Comment by bfirsh** [comment]
Thanks for the suggestion! What's the advantage of this method over using existing tools like `nohup`?

**Comment by mastertinner** [comment]
That I wouldn't have to install MongoDB and could have it run in a container instead of on my local machine. I would develop my apps locally against this MongoDB and as soon as I'm done, just stop and remove the container.

**Comment by themalkolm** [comment]
Inception! Maybe simply run it in bg then?

`$ mongo &`

**Comment by themalkolm** [comment]
No, I meant the idea of whalebrew to treat docker containers as native commands. The naive way to run native commands in bg is &.

Maybe whalebrew should have `--docker-args` somewhere (in package definition?) so such things could be easily hacked on package level without requiring to fix the binary right away.

**Comment by mastertinner** [comment]
Thanks for your feedback, @themalkolm! Since we already have some docker args covered through the labels that you can give your image (e.g. `io.whalebrew.config.environment` for the `-e` flag or `io.whalebrew.config.ports` for the `-p` flag), wouldn't it be more consequent to provide a `io.whalebrew.config.detach` label for the `-d` flag as described above? In that way, the developer of the specific whalebrew image could easily set whether their image should run as detached or not.

**Comment by nithin-bose** [comment]
So if we do make whalebrew to run a detached command, how do you then kill/stop that container?

**Comment by themalkolm** [comment]
Yep, this is the problem with detached containers I see. Now you suddenly break out of this abstraction and need to know that it is actually a docker container.

**Comment by mastertinner** [comment]
good point... you would really need to run `docker stop`...

**Comment by themalkolm** [comment]
This is why I suggested to simply run this cmd in bg so whenever it is killed it will stop the container automatically. But I think this is *so* error prone to random containers floating around.

**Comment by bfirsh** [comment]
Yup, it's a leaky abstraction, but I think commands by default _always_ running in the background is even worse, because you have no idea it is running in the background.

To solve that I suppose if it starts something in detached mode, it could say something like "Starting mongo in detached mode. Use Docker commands to manage it." 

If you run `mongo &`, it is explicit that you're running something in the background. Both `ps` and `docker ps` will work to see it is running, which is pretty neat.

**Comment by LongLiveCHIEF** [comment]
I can image there would also be advantages to being able to use the docker DNS'ing to connect to whalebrew'd containers using container names.

**Comment by billinghamj** [comment]
Effectively isn't the ideal to be able to have your brew services running inside docker containers? So e.g. my local redis, postgres, mongo, elasticsearch etc can all be running off cleanly separated docker containers?
